Error unexpected EOF occurs when filter, prefix, tags are missing in aws_s3_bucket_metric

_This issue was originally opened by @papalagichen as hashicorp/terraform#21029. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.11.13
+ provider.aws v1.60.0
+ provider.null v2.1.1
+ provider.template v2.1.1
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
resource aws_s3_bucket_metric bucket-metric-prefix-root {
  bucket = "${aws_s3_bucket.bucket.bucket}"
  name = "prefix-root"
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

### Expected Behavior
<!--
What should have happened?
-->
According to [s3_bucket_metric document](https://www.terraform.io/docs/providers/aws/r/s3_bucket_metric.html), `filter` argument is optional. `filter` should be ok to be absent. Further, `prefix` and `tags` configurations under `filter` argument are both optional. It should allow `filter` argument to be empty.

Since AWS S3 allows empty prefix / tags, it makes sense to support empty string in `prefix` and `tags` under `filter` argument.

### Actual Behavior
<!--
What actually happened?
-->
```
* aws_s3_bucket_metric.bucket-metric-prefix-root: 1 error(s) occurred:

* aws_s3_bucket_metric.bucket-metric-prefix-root: unexpected EOF 
```

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
`terraform init`
`terraform plan`
`terraform apply`

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
I have a workaround by putting `prefix = "/"` under `filter`, S3 metric filter will be created and becomes empty finally.

The Terraform command was executed directly from MacOS 10.14.4 with Bash shell 5.0.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- hashicorp/terraform#6017

-->

Turn on TF_LOG

```
panic: interface conversion: interface {} is nil, not map[string]interface {}
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: goroutine 52 [running]:
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsS3BucketMetricPut(0xc00064a230, 0x408f060, 0xc0003c0800, 0xc00064a230, 0x0)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_s3_bucket_metric.go:65 +0x624
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc0005564d0, 0xc0007440a0, 0xc000157ee0, 0x408f060, 0xc0003c0800, 0x100b601, 0xc0003dcb80, 0x10bddbc)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:225 +0x351
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc000608620, 0xc000744050, 0xc0007440a0, 0xc000157ee0, 0xc000086400, 0x18, 0x9701b00)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0x9c
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc000156d00, 0xc0001574c0, 0xc000676290, 0x0, 0x0)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:527 +0x57
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: reflect.Value.call(0xc0001741e0, 0xc0000b2d30, 0x13, 0x477008d, 0x4, 0xc0003dcf18, 0x3, 0x3, 0xc00028d800, 0xc0000b8fc0, ...)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/goenv/versions/1.11.5/src/reflect/value.go:447 +0x454
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: reflect.Value.Call(0xc0001741e0, 0xc0000b2d30, 0x13, 0xc000b71718, 0x3, 0x3, 0xc000326b60, 0xc000326b20, 0xc000326ae0)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/goenv/versions/1.11.5/src/reflect/value.go:308 +0xa4
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: net/rpc.(*service).call(0xc0003ce140, 0xc0003320a0, 0xc0004b2138, 0xc0004b2160, 0xc000192480, 0xc0003d08a0, 0x3a260c0, 0xc0001574c0, 0x16, 0x3a26100, ...)
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/goenv/versions/1.11.5/src/net/rpc/server.go:384 +0x14e
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: created by net/rpc.(*Server).ServeCodec
2019-04-16T17:29:23.269-0700 [DEBUG] plugin.terraform-provider-aws_v1.60.0_x4: 	/opt/goenv/versions/1.11.5/src/net/rpc/server.go:481 +0x47e
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalWriteState
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalIf
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalWriteState
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalWriteDiff
2019/04/16 17:29:23 [TRACE] root: eval: *terraform.EvalApplyPost
2019/04/16 17:29:23 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_s3_bucket_metric.enigma-so-graph-metric-prefix-root: unexpected EOF
2019/04/16 17:29:23 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_s3_bucket_metric.enigma-so-graph-metric-prefix-root: unexpected EOF
2019/04/16 17:29:23 [TRACE] [walkApply] Exiting eval tree: aws_s3_bucket_metric.enigma-so-graph-metric-prefix-root
```

--- a/aws/resource_aws_s3_bucket_metric.go
+++ b/aws/resource_aws_s3_bucket_metric.go
@@ -62,8 +62,9 @@ func resourceAwsS3BucketMetricPut(d *schema.ResourceData, meta interface{}) erro
 
 	if v, ok := d.GetOk("filter"); ok {
 		filterList := v.([]interface{})
-		filterMap := filterList[0].(map[string]interface{})
-		metricsConfiguration.Filter = expandS3MetricsFilter(filterMap)
+		if filterMap, ok := filterList[0].(map[string]interface{}); ok {
+			metricsConfiguration.Filter = expandS3MetricsFilter(filterMap)
+		}
 	}
 
 	input := &s3.PutBucketMetricsConfigurationInput{

--- a/aws/resource_aws_s3_bucket_metric_test.go
+++ b/aws/resource_aws_s3_bucket_metric_test.go
@@ -292,6 +292,29 @@ func TestAccAWSS3BucketMetric_basic(t *testing.T) {
 	})
 }
 
+func TestAccAWSS3BucketMetric_WithEmptyFilter(t *testing.T) {
+	var conf s3.MetricsConfiguration
+	rInt := acctest.RandInt()
+	resourceName := "aws_s3_bucket_metric.test"
+
+	bucketName := fmt.Sprintf("tf-acc-%d", rInt)
+	metricName := t.Name()
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSS3BucketMetricsConfigWithEmptyFilter(bucketName, metricName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSS3BucketMetricsConfigExists(resourceName, &conf),
+				),
+			},
+		},
+	})
+}
+
 func TestAccAWSS3BucketMetric_WithFilterPrefix(t *testing.T) {
 	var conf s3.MetricsConfiguration
 	rInt := acctest.RandInt()
@@ -602,6 +625,18 @@ resource "aws_s3_bucket" "bucket" {
   bucket = "%s"
 }
 `, name)
+}
+
+func testAccAWSS3BucketMetricsConfigWithEmptyFilter(bucketName, metricName string) string {
+	return fmt.Sprintf(`
+%s
+
+resource "aws_s3_bucket_metric" "test" {
+  bucket = "${aws_s3_bucket.bucket.id}"
+  name = "%s"
+  filter {}
+}
+`, testAccAWSS3BucketMetricsConfigBucket(bucketName), metricName)
 }
 
 func testAccAWSS3BucketMetricsConfigWithFilterPrefix(bucketName, metricName, prefix string) string {

